### PR TITLE
[bug 1225595] Migrate from Tower to Puente

### PR DESCRIPTION
**Pull request by rehandalal**
f?

### CODE REVIEW

File: kitsune/dashboards/templates/dashboards/contributors.html [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {% extends "base.html" %}
 {% from "dashboards/includes/macros.html" import print_readout, product_choice_list, print_subscription_menu, category_filter with context %}
 {% from "dashboards/includes/macros.html" import localization_sidebar_nav %}
-{% set title = _('[{product}] Knowledge Base Dashboard')|f(product=_(product.title, 'DB: products.Product.title') if product else _('All Products')) %}
+{% set title = _('[{product}] Knowledge Base Dashboard')|f(product=pgettext('DB: products.Product.title', product.title) if product else _('All Products')) %}
```
Comment (willkg) [inline kitsune/dashboards/templates/dashboards/contributors.html]: Using `|f` won't escape the product title. Are you sure the product title can never ever contain unsafe characters? Seems like the `|f` should get converted to a `|fe`.

File: kitsune/dashboards/templates/dashboards/contributors_old.html [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {% extends "base.html" %}
 {% from "dashboards/includes/macros.html" import print_readout, product_choice_list, print_subscription_menu with context %}
 {% from "dashboards/includes/macros.html" import localization_sidebar_nav %}
-{% set title = _('[{product}] Knowledge Base Dashboard')|f(product=_(product.title, 'DB: products.Product.title') if product else _('All Products')) %}
+{% set title = _('[{product}] Knowledge Base Dashboard')|f(product=pgettext('DB: products.Product.title', product.title) if product else _('All Products')) %}
```
Comment (willkg) [inline kitsune/dashboards/templates/dashboards/contributors_old.html]: Using `|f` won't escape the product title. Seems like this also should get converted to `|fe`.

File: kitsune/dashboards/templates/dashboards/localization.html [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {% extends "base.html" %}
 {% from "dashboards/includes/macros.html" import print_readout, overview_section, product_choice_list, print_subscription_menu with context %}
 {% from "dashboards/includes/macros.html" import localization_sidebar_nav %}
-{% set title = _('[{locale}][{product}] Localization Dashboard')|f(locale=current_locale, product=_(product.title, 'DB: products.Product.title') if product else _('All Products')) %}
+{% set title = _('[{locale}][{product}] Localization Dashboard')|f(locale=current_locale, product=pgettext('DB: products.Product.title', product.title) if product else _('All Products')) %}
```
Comment (willkg) [inline kitsune/dashboards/templates/dashboards/localization.html]: Using `|f` won't escape the title. Seems like this should be `|fe`, too.

File: kitsune/products/templates/products/mobile/product.html [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-{% set title = _(product.title, 'DB: products.Product.title') %}
+{% set title = pgettext('DB: products.Product.title', product.title) %}
 {% set headline = title %}
 {% set classes = 'product-landing' %}
 {% set styles = ('mobile-products',) %}
 {% set scripts = ('mobile-products',) %}
 {% set return_url = url('products') %}
 
-{% block title %}{{ _('{product} Help')|f(product=_(product.title, 'DB: products.Product.title')) }}{% endblock %}
+{% block title %}{{ _('{product} Help')|f(product=pgettext('DB: products.Product.title', product.title)) }}{% endblock %}
```
Comment (willkg) [inline kitsune/products/templates/products/mobile/product.html]: Using `|f` doesn't escape text. Seems like this should be `|fe`.

File: kitsune/search/tests/test_search_simple.py [hunk 5]
```diff
@@ -17,7 +17,7 @@
 
 
 def render(s, context):
-    t = jingo.env.from_string(s)
+    t = jingo.get_env().from_string(s)
```
Comment (willkg) [inline kitsune/search/tests/test_search_simple.py]: Looks like you upgraded to jingo 0.8.1 here. I haven't tested that with Puente and have no idea if that works. Let me know if you run into issues.

**Comment by willkg** [comment]
Running tests, `./manage.py extract` and `./manage.py merge` all fail with:

```
./manage.py test --noinput --nocapture --with-nicedots --logging-clear-handlers -s
Traceback (most recent call last):
  File "./manage.py", line 15, in <module>
    from kitsune.sumo.monkeypatch import patch  # noqa
  File "/home/willkg/mozilla/kitsune/kitsune/sumo/__init__.py", line 7, in <module>
    patch()
  File "/home/willkg/mozilla/kitsune/kitsune/sumo/monkeypatch.py", line 100, in patch
    jingo.get_env().install_gettext_translations(translation, newstyle=True)
  File "/home/willkg/.virtualenvs/kitsune/lib/python2.7/site-packages/jingo/__init__.py", line 126, in get_env
    loaders += [jinja2.PackageLoader(c.name) for c in apps.get_app_configs()]
  File "/home/willkg/.virtualenvs/kitsune/lib/python2.7/site-packages/django/apps/registry.py", line 137, in get_app_configs
    self.check_apps_ready()
  File "/home/willkg/.virtualenvs/kitsune/lib/python2.7/site-packages/django/apps/registry.py", line 124, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

Did you test this? Did you check to make sure no strings changed in the `.pot` file?

**Comment by willkg** [comment]
Do we know if pgettext evaluates in templates? Also, do we know if it escapes correctly? It's not something that comes from Puente's i18n extension and so it doesn't get marked by Jinja2's gettext code as safe.

**Comment by willkg** [comment]
My bad--I totally missed the WIP and f? and thought this was done. I apologize for being flippant.

So... my thoughts are:

1. Does this work with jingo 0.8.1? Everything is failing with jingo-related things. I don't know if that's a Puente issue or a Jingo issue.
2. Is pgettext in the template rendering context? If so, is it escaping properly? Do we know if it's going through Jinja2's gettext-y stuff?
3. We'll need to check strings because it's likely we'll have to fix at least some of them since Puente doesn't collapse whitespace in gettext/ngettext calls.

**Comment by willkg** [comment]
I did some digging. I don't see any support for msgctxt in gettext/ngettext in templates. So switching from Tower to Puente is problematic because you lose that feature. That sucks.

I'm going to open up a bug in Puente about this. I think what I'll do is write a wrapper around the Django pgettext and then stick that in the environment the i18n extension builds. Then I'll see if that works with extraction. That's the short-term fix. Long-term, I'm not sure what the options are. Maybe a Jinja2 fix.

I'll write up that issue and look into it later today. It's related to another issue I'm working no, too.

**Comment by willkg** [comment]
I pushed out Puente 0.4 that implements pgettext and npgettext, so those should work in templates now. Further, @robhudson fixed the `STANDALONE_DOMAINS` thing which would have bitten you once you got past `./manage.py extract` not running.

Hopefully that unblocks this. If not, let me know.

**Comment by willkg** [comment]
How's this going? Are you planning to try to land it before MOZLANDO?

**Comment by rehandalal** [comment]
updated ^

**Comment by mythmon** [comment]
Some weird errors in the Travis output.

**Comment by rehandalal** [comment]
@mythmon not sure what's going on now with test failures. I fixed the one issue that made sense to me.

**Comment by willkg** [comment]
Skimming test results:

```
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_badges_listed
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_bio_links_nofollow
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_deactivate_button
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_num_documents
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_view_ProfileFactory
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_view_profile_mine
ERROR: kitsune/users/tests/test_views.py:UserProfileTests.test_ProfileFactory
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_badges_listed
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_bio_links_nofollow
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_deactivate_button
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_num_documents
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_view_ProfileFactory
ERROR: kitsune/users/tests/test_templates.py:ViewProfileTests.test_view_profile_mine
ERROR: kitsune/users/tests/test_views.py:UserProfileTests.test_ProfileFactory
```

That error comes from Jinja2. Either Jingo changed which undefined gets used by default or it got changed in settings or something. Anyhow, the 'user_awards' variable needs to be added to the rendering context and then you're good to go.

These all have "list index out of range":

```
ERROR: kitsune/wiki/tests/test_templates.py:TranslateTests.test_translate_GET_with_perm
ERROR: kitsune/wiki/tests/test_templates.py:ReviewRevisionTests.test_default_significance
```

It looks like some of the form fields are getting escaped and that's probably causing the problems. That might be indicative that a lot of other things are being escaped, too.

**Comment by rehandalal** [comment]
50th times the charm ^

**Comment by willkg** [comment]
All checks passed!

**Comment by rehandalal** [comment]
woooo!

**Comment by willkg** [comment]
I did a few review passes on this and the recent commits and everything looked good as far as I could tell. I think it's worth doing a bunch of "by hand" testing especially the important parts of the site.

Has that happened, yet? If not, what's the testing plan?

**Comment by mythmon** [comment]
We can put this on dev and ask Rebecca to do some QA on the site (by NIing her in the bug). That would also give us a chance to do some manual testing ourselves.

**Comment by willkg** [comment]
I think that sounds prudent. +1 for testing on -dev.

**Comment by willkg** [comment]
this pr has the devil in it. anyone who has touched it will have seven years bad luck.

**Comment by rehandalal** [comment]
updated to fix issue found in QA ^

**Comment by mythmon** [comment]
Alright. Lets do this.

**Pull request merged**